### PR TITLE
[Model] Add support for xverse 

### PR DESCRIPTION
**Pull request by hxer7963**
Add support for xverse models.

We tested the xverse 7B/13B/65B chat models and the quantized GPTQ model locally, and vllm responses are normal. 

You can verify by downloading xverse models from Hugging Face (https://huggingface.co/xverse) or Modelscope (https://www.modelscope.cn/search?search=xverse).

Furthermore, before the PR, we executed format.sh.

Request: After the current PR is merged, can we add a new tag so that the latest version of the package supports inference of xverse models via pip installation.

### CODE REVIEW

**Comment by Alec-Lin** [comment]
What is the difference of arch between llama and xverse?

**Comment by hxer7963** [comment]
> What is the difference of arch between llama and xverse?

Nice question.

The current xverse model architecture is no different from llama, and it is expected that xverse will add moe features within two weeks and merge them into `xverse.py`.

To maintain an independent update progress, it is necessary to separately support the xverse architecture in VLLM.

**Comment by hxer7963** [comment]
Hello @WoosukKwon ,

I've submitted a PR #3610 and I would greatly appreciate your help with reviewing the code changes  and provide your feedback.

Thank you for your time and assistance.

Best regards,
willhe.

**Comment by WoosukKwon** [comment]
@hxer7963 Could you please fix the formatting error by running `pip install -r requirements-dev.txt; ./format.sh`?

File: README.md [hunk 1]
```diff
@@ -90,6 +90,7 @@ vLLM seamlessly supports many Hugging Face models, including the following archi
 - StableLM(`stabilityai/stablelm-3b-4e1t`, `stabilityai/stablelm-base-alpha-7b-v2`, etc.)
 - Starcoder2(`bigcode/starcoder2-3b`, `bigcode/starcoder2-7b`, `bigcode/starcoder2-15b`, etc.)
 - Yi (`01-ai/Yi-6B`, `01-ai/Yi-34B`, etc.)
+- Xverse (`xverse/XVERSE-7B-Chat`, `xverse/XVERSE-13B-Chat`, `xverse/XVERSE-65B-Chat`, etc.)
```
Comment (WoosukKwon) [inline README.md:93]: nit: Let's keep the alphabetic order :)
```suggestion
- Xverse (`xverse/XVERSE-7B-Chat`, `xverse/XVERSE-13B-Chat`, `xverse/XVERSE-65B-Chat`, etc.)
- Yi (`01-ai/Yi-6B`, `01-ai/Yi-34B`, etc.)
```

File: vllm/model_executor/models/xverse.py [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+from vllm.model_executor.layers.vocab_parallel_embedding import (
+    VocabParallelEmbedding, ParallelLMHead, DEFAULT_VOCAB_PADDING_SIZE)
+from vllm.model_executor.parallel_utils.parallel_state import (
+    get_tensor_model_parallel_world_size)
+from vllm.model_executor.sampling_metadata import SamplingMetadata
+from vllm.model_executor.weight_utils import (default_weight_loader,
+                                              hf_model_weights_iterator)
+from vllm.sequence import SamplerOutput
+
+KVCache = Tuple[torch.Tensor, torch.Tensor]
```
Comment (WoosukKwon) [inline vllm/model_executor/models/xverse.py:49]: Please replace `KVCache` with `torch.Tensor`. We changed the type recently.

**Comment by esmeetu** [comment]
Can we simply use this: `"XverseForCausalLM": ("llama", "LlamaForCausalLM")`?

**Comment by hxer7963** [comment]
> Can we simply use this: `"XverseForCausalLM": ("llama", "LlamaForCausalLM")`? Will the new moe model use the same name with `XverseForCausalLM`?
Yes, the new moe model will use the same name with `XverseForCausalLM`, so we want to add a new name.

File: vllm/model_executor/models/xverse.py (hunk 2 above)
Comment (hxer7963) [reply vllm/model_executor/models/xverse.py]: Thank you for WoosukKwon's reply and suggestions. The issues mentioned in the review have been fixed in the latest commit.

**Comment by hxer7963** [comment]
> Can we simply use this: `"XverseForCausalLM": ("llama", "LlamaForCausalLM")`? Will the new moe model use the same name with `XverseForCausalLM`?

Yes, the new moe model will use the same name with XverseForCausalLM, so we want to add a new name.

**Comment by hxer7963** [comment]
Hello @WoosukKwon ,

I've addressed all the feedback in the PR, and all checks have passed. 

Could you please take another look at the code changes and provide your review?

Thank you for your time and assistance.

Best regards,

hxer7963.

**Review by WoosukKwon — APPROVED** [review]
LGTM! Thanks for submitting the PR. Looking forward to the next model release!

**Pull request merged**